### PR TITLE
[OSGi] - Cleanup BudleActivator and Service registration, Update bnd and osgi deps

### PR DESCRIPTION
**Pull request by stbischof**
- cleanup BudleActivator
- cleanup Service registration
- use declarative services
- update bnd 
- update osgi deps
- base it-test

### CODE REVIEW

**Comment by stbischof** [comment]
> changes breaks everything without a clear reason.

this commit https://github.com/vaadin/flow/pull/9361/commits/06ed6c7ab69c9c236592107e05b417b3ba4db9b9 fixes the errors.

iIadded some more OSGi-related atomic commits. Most removes BudleActivator and/or bundleContect::registerService. I not uses declarative Services via @Component

**Comment by denis-anisimov** [comment]
> > changes breaks everything without a clear reason.
> 
> @denis-anisimov this commit [06ed6c7](https://github.com/vaadin/flow/commit/06ed6c7ab69c9c236592107e05b417b3ba4db9b9) fixes the errors.
> 
This is not the problem which I referred to.
It allows to run OSGi related unit tests in Eclipse.
And this is no-op if you run them via maven: maven config is done in the way which excludes OSGi dependencies from test dependencies. 
Eclipse is not able to apply maven config correctly and the tests fail.

But something is broken when I applied changes (some changes) from you PR for the `feature`  branch to the `master`  : when I run our OSGi ITs the HTTP functionality doesn't work.

The current state of my PR makes everything work. I would like to finish this endless journey and may be keep it as is in the working sate.
I will apply your PR locally and check whether HTTP still works in our ITs or not if I will have time.

But I would may be consider this as a separate extra PR : I really don't see any end of the changes. There are many many things which can be done to improve OSGi integration and cherry pick PR is not about every improvement: it should just enable initial OSGi support.

**Comment by stbischof** [comment]
i think updating bnd and osgi-deps you be a great improvement.

but i am a still not sure what OSGi related unit tests in Eclipse does not run.

**Comment by denis-anisimov** [comment]
> i think updating bnd and osgi-deps you be a great improvement.
> 
> but i am a still not sure what OSGi related unit tests in Eclipse does not run. what does i have to do to force this tests

every test which has `@RunWith(EnableOSGiRunner.class)`.

**Comment by stbischof** [comment]
and how to start? 
They run on my side after the fix
` .make().load(this, ClassLoadingStrategy.Default.CHILD_FIRST.allowExistingTypes())`

**Comment by denis-anisimov** [comment]
> and how to start? I cant not produce the error.
> They run on my side after the fix
> ` .make().load(this, ClassLoadingStrategy.Default.CHILD_FIRST.allowExistingTypes())`

I think there is a misunderstanding.........

There are no any problems with these tests.
They didn't work in Eclipse for me without this PR. I have not checked actually this PR but I guess that should fix it.
But it doesn't matter whether they work in Eclipse or not: it's rather an Eclipse issue.

The tests work even without these changes if they are executed via Maven.

And this wasn't a problem for me.

I was not able to run OSGi ITs with your latest changes for bnd version and `VaadinBundleTracker`  : that's the problem.
I haven't checked whether it's still the problem or not with this PR. I've been busy with other things.

I will check and let you know.

**Comment by stbischof** [comment]
hope this PR helps as i mentioned, osgi R7 deps and bnd are most importent

File: flow-polymer-template/src/main/java/com/vaadin/flow/component/polymertemplate/rpc/PolymerPublishedEventRpcHandler.java [hunk 1]
```diff
@@ -31,8 +32,11 @@
 /**
  * Polymer utilitiy class for handling polymer rpc events for @EventHandler.
  *
+ * Registers {@link PolymerPublishedEventRpcHandler} as a service to make it
+ * available in {@link Lookup} in an OSGi container. *
  * @since
  */
+@org.osgi.service.component.annotations.Component(immediate = true)
```
Comment (denis-anisimov) [inline flow-polymer-template/src/main/java/com/vaadin/flow/component/polymertemplate/rpc/PolymerPublishedEventRpcHandler.java:39]: No need to use FQN. Just add an import.

**Review by denis-anisimov — APPROVED** [review]

**Comment by denis-anisimov** [comment]
I confirm that nothing is broken with this PR.
Let's merge it.

Thanks!

**Comment by denis-anisimov** [comment]
@stbischof , I've reverted `bnd` `5.2.0` version change.

Bnd maven plugin is quite seriously broken

I'm packing `flow-tests/test-root-context` module with `ui` classifier as a JAR
(this module contains classes for our ITs and it's packaged as a WAR by default, but there are two classifiers: ui and test-jar).

Here are the snippets of MANIFEST with this version :
```
Import-Package: afu.org.checkerframework.checker.formatter.qual,afu.or
 g.checkerframework.checker.nullness.qual,afu.org.checkerframework.che
 cker.regex.qual,afu.org.checkerframework.checker.units.qual,afu.org.c
 heckerframework.dataflow.qual,afu.org.checkerframework.framework.qual
 ,android.net,android.net.http,android.os,android.security,android.uti
 l,ch.qos.cal10n,com.android.org.conscrypt,com.github.luben.zstd,com.g
 oogle.appengine.api,com.google.apphosting.api,com.google.gson,com.goo
 gle.gwt.core.client,com.google.gwt.xhr.client,com.sun.enterprise.web.
 connector.grizzly.comet,com.sun.grizzly.comet,com.sun.grizzly.tcp,com
 .sun.grizzly.util.buf,com.sun.grizzly.websockets,com.sun.jdi,com.sun.
 jdi.connect,com.sun.jdi.event,com.sun.jdi.request,dalvik.system,java.
 applet,java.awt,java.awt.datatransfer,java.awt.event,java.awt.image,j
 ava.beans,java.io,java.lang,java.lang.annotation,java.lang.instrument
 ,java.lang.invoke,java.lang.management,java.lang.ref,java.lang.reflec
 t,java.math,java.net,java.nio,java.nio.channels,java.nio.charset,java
 .nio.file,java.nio.file.attribute,java.nio.file.spi,java.security,jav
 a.security.cert,java.security.spec,java.sql,java.text,java.time,java.
 time.chrono,java.time.format,java.time.temporal,java.time.zone,java.u
 til,java.util.concurrent,java.util.concurrent.atomic,java.util.concur
 rent.locks,java.util.function,java.util.jar,java.util.logging,java.ut
 il.prefs,java.util.regex,java.util.stream,java.util.zip,javax.crypto,
 javax.crypto.spec,javax.enterprise.context,javax.enterprise.inject,ja
 vax.imageio,javax.inject,javax.jnlp,javax.lang.model.element,javax.na
 ming,javax.naming.directory,javax.naming.ldap,javax.net,javax.net.ssl
 ,javax.portlet,javax.script,javax.security.auth.x500,javax.servlet;ve
 rsion="[3.1,4)",javax.servlet.annotation;version="[3.1,4)",javax.serv
 let.http;version="[3.1,4)",javax.swing,javax.swing.border,javax.swing
 .event,javax.swing.filechooser,javax.swing.table,javax.swing.text,jav
 ax.swing.tree,javax.websocket,javax.websocket.server,javax.xml.dataty
 pe,javax.xml.namespace,javax.xml.parsers,javax.xml.stream,javax.xml.t
 ransform,javax.xml.transform.dom,javax.xml.transform.sax,javax.xml.tr
 ansform.stream,javax.xml.validation,javax.xml.xpath,jdk.internal.misc
 ,org.antlr.v4.gui,org.apache.avalon.framework.logger,org.apache.catal
 ina,org.apache.catalina.comet,org.apache.catalina.connector,org.apach
 e.catalina.util,org.apache.catalina.websocket,org.apache.commons.lang
 .text,org.apache.commons.pool2,org.apache.commons.pool2.impl,org.apac
 he.coyote.http11.upgrade,org.apache.harmony.xnet.provider.jsse,org.ap
 ache.log,org.apache.log4j,org.apache.shiro,org.apache.shiro.mgt,org.a
 pache.shiro.subject,org.apache.shiro.web.env,org.apache.shiro.web.sub
 ject,org.apache.shiro.web.util,org.apache.xmlbeans,org.atmosphere.jbo
 ss.as.websockets,org.atmosphere.jboss.as.websockets.servlet,org.atmos
 phere.jboss.websockets,org.atmosphere.jboss.websockets.frame,org.atmo
 sphere.jersey.util,org.brotli.dec,org.checkerframework.checker.format
 ter.qual,org.checkerframework.checker.nullness.qual,org.checkerframew
 ork.checker.regex.qual,org.checkerframework.checker.units.qual,org.ch
 eckerframework.dataflow.qual,org.checkerframework.framework.qual,org.
 conscrypt,org.eclipse.jetty.continuation,org.eclipse.jetty.websocket,
 org.eclipse.jetty.websocket.api,org.eclipse.jetty.websocket.server,or
 g.eclipse.jetty.websocket.servlet,org.glassfish.grizzly,org.glassfish
 .grizzly.comet,org.glassfish.grizzly.http.server,org.glassfish.grizzl
 y.websockets,org.glassfish.hk2.osgiresourcelocator,org.ietf.jgss,org.
 jboss.servlet.http,org.jboss.vfs,org.joda.convert,org.json,org.junit,
 org.junit.internal,org.junit.rules,org.junit.runner,org.junit.runners
 ,org.junit.runners.model,org.mortbay.util.ajax,org.osgi.framework;ver
 sion="[1.9,2)",org.osgi.framework.wiring;version="[1.2,2)",org.osgi.s
 ervice.http;version="[1.2,2)",org.osgi.util.tracker;version="[1.5,2)"
 ,org.springframework.beans.factory.annotation,org.springframework.boo
 t.autoconfigure.condition,org.springframework.boot.autoconfigure.jack
 son,org.springframework.boot.autoconfigure.web.servlet,org.springfram
 ework.boot.context.properties,org.springframework.context,org.springf
 ramework.context.annotation,org.springframework.http,org.springframew
 ork.http.converter.json,org.springframework.stereotype,org.springfram
 ework.util,org.springframework.web.bind.annotation,org.springframewor
 k.web.servlet.mvc.condition,org.springframework.web.servlet.mvc.metho
 d,org.springframework.web.servlet.mvc.method.annotation,org.testng,or
 g.tukaani.xz,org.w3c.dom,org.w3c.dom.bootstrap,org.w3c.dom.ls,org.xml
 .sax,org.xml.sax.ext,org.xml.sax.helpers,sun.misc,sun.security.ssl,su
 n.security.util,weblogic.websocket,weblogic.websocket.annotation
```

```

Service-Component: OSGI-INF/com.vaadin.flow.osgi.Activator.xml,OSGI-IN
 F/com.vaadin.flow.server.osgi.OSGiResourceProvider.xml,OSGI-INF/com.v
 aadin.flow.server.osgi.VaadinBundleTracker.xml,OSGI-INF/com.vaadin.fl
 ow.component.polymertemplate.rpc.PolymerPublishedEventRpcHandler.xml
Bundle-Developers: vaadin;organization=Vaadin;organizationUrl="http://
 vaadin.com"
Provide-Capability: osgi.service;objectClass:List<String>="com.vaadin.
 flow.component.template.internal.DeprecatedPolymerPublishedEventHandl
 er";uses:="com.vaadin.flow.component.template.internal",osgi.service;
 objectClass:List<String>="com.vaadin.flow.di.ResourceProvider";uses:=
 "com.vaadin.flow.di"
```

`afu.org`, `android.util`.... : I have no any idea where it comes from .
This is absolutely wrong.

This bundle doesn't contain service components except one `Activator`. All other components comes from dependencies.
And it doesn't provide ant capability. 

Many manifest headers are wrong.

Here is the manifest made by `3.3.0`  version:
```
Import-Package: com.vaadin.flow.component;version="[6.0,7)",com.vaadin
 .flow.component.dependency;version="[6.0,7)",com.vaadin.flow.componen
 t.dnd;version="[6.0,7)",com.vaadin.flow.component.html;version="[6.0,
 7)",com.vaadin.flow.component.internal;version="[6.0,7)",com.vaadin.f
 low.component.littemplate;version="[6.0,7)",com.vaadin.flow.component
 .page;version="[6.0,7)",com.vaadin.flow.component.polymertemplate;ver
 sion="[6.0,7)",com.vaadin.flow.component.template;version="[6.0,7)",c
 om.vaadin.flow.data.value;version="[6.0,7)",com.vaadin.flow.dom;versi
 on="[6.0,7)",com.vaadin.flow.function;version="[6.0,7)",com.vaadin.fl
 ow.i18n;version="[6.0,7)",com.vaadin.flow.internal;version="[6.0,7)",
 com.vaadin.flow.internal.nodefeature;version="[6.0,7)",com.vaadin.flo
 w.router;version="[6.0,7)",com.vaadin.flow.server;version="[6.0,7)",c
 om.vaadin.flow.server.communication;version="[6.0,7)",com.vaadin.flow
 .shared;version="[6.0,7)",com.vaadin.flow.shared.communication;versio
 n="[6.0,7)",com.vaadin.flow.shared.ui;version="[6.0,7)",com.vaadin.fl
 ow.templatemodel;version="[6.0,7)",elemental.json;version="[2.8,3)",j
 avax.servlet;version="[3.1,4)",javax.servlet.annotation;version="[3.1
 ,4)",javax.servlet.http;version="[3.1,4)",org.apache.commons.io;versi
 on="[1.4,2)",org.apache.commons.lang3;version="[3.9,4)",org.osgi.fram
 ework;version="[1.9,2)",org.osgi.service.http;version="[1.2,2)",org.o
 sgi.util.tracker;version="[1.5,2)",org.slf4j;version="[1.7,2)"
Require-Capability: osgi.ee;filter:="(&(osgi.ee=JavaSE)(version=1.8))"
Service-Component: OSGI-INF/com.vaadin.flow.osgi.Activator.xml
```

The remaining part contains just bundle name and `Export-Package` section.
So the manifest is much shorter than the one generated with `5.2.0` .
And this content is correct.

So `5.2.0`  bnd is seriously broken and we may not have this version.

**Comment by stbischof** [comment]
@maho7791
could you please helpe me here. I think using bnd 5.2 should help us more then 3.3

**Comment by denis-anisimov** [comment]
Also, I'm not sure whether this is related or not but 
```
@Activate
public VaadinBundleTracker(BundleContext context)
```

doesn't work in our ITs config: OSGi tries to instantiate the component and it wants to use a default cosntructor.
But there is no default contractor and it fails to create and register the service.
`@Activate` annotation should be used properly with compendium `7`  but in runtime it's apparently broken.
I'm not sure whether this is related to the compendium version or `bnd` version.
I had to revert your changes related to `VaadinBundleTracker` as well.

It's weird why it's broken because I've checked that it worked before accept the PR.

I know that I didn't rebuild the ITs module from scratch to test but used the old one. Probably that's the reason.
But now the code doesn't work even with the correct IT module: the same error is shown in the console about absence of default constructor.

**Comment by denis-anisimov** [comment]
> 
> @denis-anisimov These better Manifest helps the resolver a lot to resove a working framework and ServiceLoader handling!
> 
> ```
> Service-Component: OSGI-INF/com.vaadin.flow.osgi.Activator.xml,OSGI-IN
>  F/com.vaadin.flow.server.osgi.OSGiResourceProvider.xml,OSGI-INF/com.v
>  aadin.flow.server.osgi.VaadinBundleTracker.xml,OSGI-INF/com.vaadin.fl
>  ow.component.polymertemplate.rpc.PolymerPublishedEventRpcHandler.xml
> Bundle-Developers: vaadin;organization=Vaadin;organizationUrl="http://
>  vaadin.com"
> Provide-Capability: osgi.service;objectClass:List<String>="com.vaadin.
>  flow.component.template.internal.DeprecatedPolymerPublishedEventHandl
>  er";uses:="com.vaadin.flow.component.template.internal",osgi.service;
>  objectClass:List<String>="com.vaadin.flow.di.ResourceProvider";uses:=
>  "com.vaadin.flow.di"
> ```

I don't understand what you mean.
As I said: `flow-tests/test-root-context` doesn't provide any capability.
So `Provide-Capability` header should not be in its manifest at all.

One more time: `flow-tests/test-root-context` .
`DeprecatedPolymerPublishedEventHandler`  capability is provided by `flow-polymer-template` bundle.
`ResourceProvider` capability is provided by the `flow-server` bundle.

 `flow-tests/test-root-context`  doesn't provide _ANY_ capability. Don't see how this incorrect info may help to the resolver.

**Comment by stbischof** [comment]
```
@Activate
public VaadinBundleTracker(BundleContext context)
```

Injecting bundleContext into a Constructor kame with R7 . bnd (with the newer versions) generates a xml file that describes this
```
Service-Component: [...] OSGI-INF/com.v
 aadin.flow.server.osgi.VaadinBundleTracker.xml, [...] 
```

**Comment by juergen-albert** [comment]
> @juergen-albert
> could you please helpe me here. I think using bnd 5.2 should help us more then 3.3.

I haven't used maven for quite some time, but as far as I remember does the war packaging cause maven to pack all the defined dependencies that are in certain scopes in the resulting war. This could mean, that bnd is analyzing all the dependent artifacts, when asked to build a manifest for it.

@denis-anisimov can you provide a jar with this large manifest, so we can have a look?

**Comment by juergen-albert** [comment]
Besides this: Please use the latest bnd, because it produces much better metadata!

**Comment by stbischof** [comment]
https://gist.github.com/stbischof/206dfb270d5dcecae350db999402d7be

**Comment by juergen-albert** [comment]
As I have guessed. The dependencies are part of the jars classpath. 
This is not really BNDs fault. Maven adds the all dependencies to the containers classpath, because packaging is war. BND then takes the classpath and handles it as it should. Version 3.3 simply did not properly handle jar-classpath libs.

The question would be, what the idea behind this jars is. Should they even be bundles or even jars? What is the idea behind packaging everything including the kitchen sink in the jars?

**Comment by stbischof** [comment]
Vaadin_checker

**Comment by stbischof** [comment]
if i exclude   `org.checkerframework - org.checkerframework` in the pom file the 
```
<dependency>
            <groupId>com.vaadin</groupId>
            <artifactId>flow-html-components-testbench</artifactId>
            <version>${project.version}</version>
            <scope>test</scope>
            <exclusions>
                <exclusion>
                    <groupId>org.checkerframework</groupId>
                    <artifactId>checker-compat-qual</artifactId>
                </exclusion>
            </exclusions>
        </dependency>
```

- `flow-test-root-context-npm-6.0-SNAPSHOT.war` does not have the entry `WEB-INF/lib/checker-compat-Qual-2.0.0.jar`
- f`low-test-root-context-npm-6.0-SNAPSHOT-ui.jar` does not have the Import-Package Entry `afu.org.checkerframework.checker.`


now:
```
Import-Package: ch.qos.cal10n,com.github.luben.zstd,com.google.appengine
 .api,com.google.apphosting.api,com.sun.enterprise.web.connector.grizzly
 .comet,com.sun.grizzly.comet,com.sun.grizzly.tcp,com.sun.grizzly.util.b
 uf,com.sun.grizzly.websockets,com.sun.jdi,com.sun.jdi.connect,com.sun.j
 di.e
```

**Comment by stbischof** [comment]
> ```
> @Activate
> public VaadinBundleTracker(BundleContext context)
> ```
> doesn't work in our ITs config: OSGi tries to instantiate the component and it wants to use a default cosntructor.
> But there is no default contractor and it fails to create and register the service.
> `@Activate` annotation should be used properly with compendium 7 but in runtime it's apparently broken.

**Comment by stbischof** [comment]
when i build:

first time failed with this:
```
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 3.233 s - in com.vaadin.flow.uitest.ui.template.ExceptionsDuringPropertyUpdatesIT
[INFO] 
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   InternalErrorIT.enableSessionExpiredNotification_sessionExpired_notificationShown:92 After enabling the 'Session Expired' notification, the page should not be refreshed after killing the session
[INFO] 
[ERROR] Tests run: 310, Failures: 1, Errors: 0, Skipped: 10
```

second time it workes!
here is the full log of a build.
https://gist.github.com/stbischof/f0ac323dc3ffdf5eb8597a39e6059f47

**Comment by stbischof** [comment]
I added `waitUntil(driver -> isMessageUpdated());` after the `clickButton(UPDATE);`

I run the test of `flow-tests/test-root-context` 7 of 8 Times  without errors. 
commit https://github.com/bipolis/flow/commit/a6b7ab4c8c1be243f7cf2148094ca4a286d2bf61

the one error happened here (completely other test):
```
[ERROR] Failures: 
[ERROR]   DomEventFilterIT.debounce:64->assertMessages:174 Unexpected message at index 0 expected:<T[hrottle: ab]> but was:<T[railing: a]>
[INFO] 
[ERROR] Tests run: 310, Failures: 1, Errors: 0, Skipped: 10
```

**Comment by denis-anisimov** [comment]
> As I have guessed. The dependencies are part of the jars classpath.
> This is not really BNDs fault. Maven adds the all dependencies to the containers classpath, because packaging is war. BND then takes the classpath and handles it as it should. Version 3.3 simply did not properly handle jar-classpath libs.
> 
> The question would be, what the idea behind this jars is. Should they even be bundles or even jars? What is the idea behind packaging everything including the kitchen sink in the jars?

Well, any way is good for me.

bnd maven plugin version `3.3.0`  doesn't put extra things in manifest.
As I mentioned : the module is normally a WAR which contains UI views for ITs. And it's executed being deployed to jetty.
I don't want to duplicate the same classes in another module.
So the same classes are packaged as JAR via jar plugin and classifier ui:
```
<plugin>
                <groupId>org.apache.maven.plugins</groupId>
                <artifactId>maven-jar-plugin</artifactId>
                <configuration>
                    <archive>
                        <manifestFile>${project.build.outputDirectory}/META-INF/MANIFEST.MF</manifestFile>
                        <index>false</index>
                        <manifest>
                            <addDefaultImplementationEntries>true</addDefaultImplementationEntries>
                        </manifest>
                    </archive>
                </configuration>
                <executions>
                    <execution>
                        <id>ui-tests</id>
                        <goals>
                            <goal>test-jar</goal>
                        </goals>
                    </execution>
                    <execution>
                        <id>ui</id>
                        <goals>
                            <goal>jar</goal>
                        </goals>
                        <configuration>
                            <classifier>ui</classifier>
                        </configuration>
                    </execution>
                </executions>
            </plugin>
```
And the module has bnd file for bundle manifest.
`3.3.0` bnd maven plugin doesn't add extra imports into the manifest. 
`5.2.0` bnd maven plugin puts everything.
So it's related to the version for sure.

If you know another way of proper packaging one module as a WAR and as a WAB please let me know.
I will just use it: there are no any problems with this.

**Comment by denis-anisimov** [comment]
@stbischof , let's don't discuss things here in the closed PR.
I just let you know the issue.

So: for now bnd version `3.3.0` will be used.
I've reverted almost all changes from this PR.

I will not accept any changes into the #9321 until it's merged and until `osgi` repo is configured for validation, sorry.
This is the infinite journey and I may not proceed with this without any validation.
Once validation is done you may create a PR against the `master` branch in Flow or in `osgi` and it will prove that nothing is broken.

**Comment by stbischof** [comment]
Sure. I will create a PR against master.

**Comment by juergen-albert** [comment]
@denis-anisimov Please refrain from using bnd 3.3.0 it is ancient and 5.2 will produce much better Manifests. I don't know your exact setup and how the WAR and WAP will be deployed in OSGi. In short: Should this be a Bundle?

If No: Just remove this line: `<manifestFile>${project.build.outputDirectory}/META-INF/MANIFEST.MF</manifestFile>`

If Yes: Use 3.3 only for this project or we might be able to tell bnd to not scan the included libs.

**Comment by stbischof** [comment]
> If you know another way of proper packaging one module as a WAR and as a WAB please let me know.
> I will just use it: there are no any problems with this.

from  bnd-doc [as above described](https://bnd.bndtools.org/md/805-instructions.html)
```
Defines the internal bundle class path, is taken into accont by bnd. That is, classes will be analyzed according to this path. The files/directories on the Bundle-ClassPath must be present in the bundle. Use Include-Resource to include these jars/directories in your bundle. In general you should not use Bundle-ClassPath since it makes things more complicated than necessary.
```

**Comment by denis-anisimov** [comment]
> @denis-anisimov Please refrain from using bnd 3.3.0 it is ancient and 5.2 will produce much better Manifests. I don't know your exact setup and how the WAR and WAP will be deployed in OSGi. In short: Should this be a Bundle?
> 
> If No: Just remove this line: `<manifestFile>${project.build.outputDirectory}/META-INF/MANIFEST.MF</manifestFile>`
> 
> If Yes: Use 3.3 only for this project or we might be able to tell bnd to not scan the included libs.

It should be a bundle, yes: that's the target . It should be possible to deploy it to the OSGi container.
The jar should be a WAB.

**Comment by stbischof** [comment]
@denis-anisimov could you please tell me which jars you want to have in the lib folder of the war.


```
activation-1.1.jar
animal-sniffer-annotations-1.14.jar
atmosphere-runtime-2.4.30.slf4jvaadin1.jar
btf-1.2.jar
byte-buddy-1.10.9.jar
checker-compat-qual-2.0.0.jar
commonmark-0.9.0.jar
commons-cli-1.2.jar
commons-codec-1.11.jar
commons-compress-1.19.jar
commons-exec-1.3.jar
commons-fileupload-1.3.3.jar
commons-io-2.5.jar
commons-lang3-3.9.jar
commons-logging-1.2.jar
error_prone_annotations-2.1.3.jar
flow-client-6.0-SNAPSHOT.jar
flow-data-6.0-SNAPSHOT.jar
flow-dnd-6.0-SNAPSHOT.jar
flow-html-components-6.0-SNAPSHOT.jar
flow-html-components-testbench-6.0-SNAPSHOT.jar
flow-lit-template-6.0-SNAPSHOT.jar
flow-polymer-template-6.0-SNAPSHOT.jar
flow-push-6.0-SNAPSHOT.jar
flow-server-6.0-SNAPSHOT.jar
flow-test-common-6.0-SNAPSHOT.jar
gentyref-1.2.0.vaadin1.jar
guava-25.0-jre.jar
gwt-elemental-2.8.2.jar
gwt-elemental-2.8.2.vaadin2.jar
handlebars-4.0.6.jar
httpclient-4.5.13.jar
httpcore-4.4.13.jar
j2objc-annotations-1.1.jar
jackson-annotations-2.10.2.jar
jackson-core-2.10.2.jar
jackson-coreutils-1.6.jar
jackson-databind-2.10.2.jar
jackson-dataformat-yaml-2.9.5.jar
jackson-datatype-jsr310-2.10.2.jar
javaparser-core-3.15.1.jar
javaparser-symbol-solver-core-3.15.1.jar
javaparser-symbol-solver-logic-3.15.1.jar
javaparser-symbol-solver-model-3.15.1.jar
javassist-3.18.1-GA.jar
javax.annotation-api-1.3.2.jar
jaxb-api-2.3.0.jar
jmustache-1.12.jar
joda-time-2.9.7.jar
jopt-simple-5.0.3.jar
json-patch-1.6.jar
json-schema-core-1.2.8.jar
json-schema-validator-2.2.8.jar
jsoup-1.12.1.jar
jsr305-3.0.1.jar
libphonenumber-8.0.0.jar
license-checker-1.0.5.jar
mailapi-1.4.3.jar
mobile-drag-drop-2.3.0-rc.1.jar
msg-simple-1.1.jar
okhttp-3.11.0.jar
okio-1.14.0.jar
ph-commons-9.1.2.jar
ph-css-6.1.1.jar
rhino-1.7R4.jar
selenium-api-3.141.59.jar
selenium-chrome-driver-3.141.59.jar
selenium-edge-driver-3.141.59.jar
selenium-firefox-driver-3.141.59.jar
selenium-ie-driver-3.141.59.jar
selenium-java-3.141.59.jar
selenium-opera-driver-3.141.59.jar
selenium-remote-driver-3.141.59.jar
selenium-safari-driver-3.141.59.jar
selenium-support-3.141.59.jar
slf4j-api-1.7.30.jar
slf4j-ext-1.7.25.jar
slf4j-simple-1.7.30.jar
snakeyaml-1.18.jar
swagger-annotations-1.5.21.jar
swagger-annotations-2.0.4.jar
swagger-codegen-2.3.1.jar
swagger-codegen-3.0.2.jar
swagger-codegen-generators-1.0.2.jar
swagger-compat-spec-parser-1.0.33.jar
swagger-core-1.5.21.jar
swagger-core-2.0.4.jar
swagger-models-1.5.21.jar
swagger-models-2.0.4.jar
swagger-parser-1.0.39.jar
swagger-parser-2.0.5.jar
swagger-parser-core-2.0.5.jar
swagger-parser-v2-converter-2.0.5.jar
swagger-parser-v3-2.0.5.jar
uri-template-0.9.jar
vaadin-testbench-core-6.3.0.jar
vaadin__vaadin-mobile-drag-drop-1.0.0.jar
validation-api-2.0.1.Final.jar
```

**Comment by denis-anisimov** [comment]
> @denis-anisimov could you please tell me which jars you want to have in the lib folder of the war.
> 
> ```
> activation-1.1.jar
> animal-sniffer-annotations-1.14.jar
> atmosphere-runtime-2.4.30.slf4jvaadin1.jar
> btf-1.2.jar
> byte-buddy-1.10.9.jar
> checker-compat-qual-2.0.0.jar
> commonmark-0.9.0.jar
> commons-cli-1.2.jar
> commons-codec-1.11.jar
> commons-compress-1.19.jar
> commons-exec-1.3.jar
> commons-fileupload-1.3.3.jar
> commons-io-2.5.jar
> commons-lang3-3.9.jar
> commons-logging-1.2.jar
> error_prone_annotations-2.1.3.jar
> flow-client-6.0-SNAPSHOT.jar
> flow-data-6.0-SNAPSHOT.jar
> flow-dnd-6.0-SNAPSHOT.jar
> flow-html-components-6.0-SNAPSHOT.jar
> flow-html-components-testbench-6.0-SNAPSHOT.jar
> flow-lit-template-6.0-SNAPSHOT.jar
> flow-polymer-template-6.0-SNAPSHOT.jar
> flow-push-6.0-SNAPSHOT.jar
> flow-server-6.0-SNAPSHOT.jar
> flow-test-common-6.0-SNAPSHOT.jar
> gentyref-1.2.0.vaadin1.jar
> guava-25.0-jre.jar
> gwt-elemental-2.8.2.jar
> gwt-elemental-2.8.2.vaadin2.jar
> handlebars-4.0.6.jar
> httpclient-4.5.13.jar
> httpcore-4.4.13.jar
> j2objc-annotations-1.1.jar
> jackson-annotations-2.10.2.jar
> jackson-core-2.10.2.jar
> jackson-coreutils-1.6.jar
> jackson-databind-2.10.2.jar
> jackson-dataformat-yaml-2.9.5.jar
> jackson-datatype-jsr310-2.10.2.jar
> javaparser-core-3.15.1.jar
> javaparser-symbol-solver-core-3.15.1.jar
> javaparser-symbol-solver-logic-3.15.1.jar
> javaparser-symbol-solver-model-3.15.1.jar
> javassist-3.18.1-GA.jar
> javax.annotation-api-1.3.2.jar
> jaxb-api-2.3.0.jar
> jmustache-1.12.jar
> joda-time-2.9.7.jar
> jopt-simple-5.0.3.jar
> json-patch-1.6.jar
> json-schema-core-1.2.8.jar
> json-schema-validator-2.2.8.jar
> jsoup-1.12.1.jar
> jsr305-3.0.1.jar
> libphonenumber-8.0.0.jar
> license-checker-1.0.5.jar
> mailapi-1.4.3.jar
> mobile-drag-drop-2.3.0-rc.1.jar
> msg-simple-1.1.jar
> okhttp-3.11.0.jar
> okio-1.14.0.jar
> ph-commons-9.1.2.jar
> ph-css-6.1.1.jar
> rhino-1.7R4.jar
> selenium-api-3.141.59.jar
> selenium-chrome-driver-3.141.59.jar
> selenium-edge-driver-3.141.59.jar
> selenium-firefox-driver-3.141.59.jar
> selenium-ie-driver-3.141.59.jar
> selenium-java-3.141.59.jar
> selenium-opera-driver-3.141.59.jar
> selenium-remote-driver-3.141.59.jar
> selenium-safari-driver-3.141.59.jar
> selenium-support-3.141.59.jar
> slf4j-api-1.7.30.jar
> slf4j-ext-1.7.25.jar
> slf4j-simple-1.7.30.jar
> snakeyaml-1.18.jar
> swagger-annotations-1.5.21.jar
> swagger-annotations-2.0.4.jar
> swagger-codegen-2.3.1.jar
> swagger-codegen-3.0.2.jar
> swagger-codegen-generators-1.0.2.jar
> swagger-compat-spec-parser-1.0.33.jar
> swagger-core-1.5.21.jar
> swagger-core-2.0.4.jar
> swagger-models-1.5.21.jar
> swagger-models-2.0.4.jar
> swagger-parser-1.0.39.jar
> swagger-parser-2.0.5.jar
> swagger-parser-core-2.0.5.jar
> swagger-parser-v2-converter-2.0.5.jar
> swagger-parser-v3-2.0.5.jar
> uri-template-0.9.jar
> vaadin-testbench-core-6.3.0.jar
> vaadin__vaadin-mobile-drag-drop-1.0.0.jar
> validation-api-2.0.1.Final.jar
> ```

Complicated question.

First of all : why do you talk about WAR?
We don't need WAR in OSGi.

So is this question really about WAR or it's about JAR for OSGi ?

**Comment by stbischof** [comment]
> It should be a bundle, yes: that's the target . It should be possible to deploy it to the OSGi container.
> The jar should be a WAB.

I would create a WAB. One single zip file with the file-extention=war and an osgi manifest.
If you create a `<packaging>war</packaging>` and use a newer `bnd-plugin` it will generate the `manifest` and includes all NON-Optional Artefacts that are Not-Test and Not-Provides Scope into the Bundle-Classpath and lib folder. Then bnd searches in all that classes for imports.

I will make a PR. Then you could comment which artefacts should be excluded.

**Pull request merged**